orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,503 @@
+				bf.Logger.Debugf("Endpoint Shuffled. Current Endpoint %s", endpoint[0].Endpoint)
+				bf.currentEndpoint = endpoint[0]
+				break
+			}
+			currentE = endpoint[0]
+		}
+	} else {
+		bf.currentEndpoint = bf.Config.Endpoints[0]
+	}
+
+	// Update current blocksource
+	bf.setBlockSource(bf.currentEndpoint)
+	bf.LastShuffledAt = bf.TimeNow()
+	return nil
+}
+
+func (bf *BlockFetcher) isShuffleTime() bool {
+	if bf.ShuffleTimeout == time.Duration(0) {
+		// It should return only false if Time based Shuffling is not enabled
+		bf.Logger.Debugf("Shuffle Disabled.")
+		return false
+	}
+	shuffle := time.Now().After(bf.LastShuffledAt.Add(bf.ShuffleTimeout))
+	bf.Logger.Debugf("Shuffle Enabled. Time to shuffle %v", shuffle)
+	return shuffle
+}
+
+func (bf *BlockFetcher) isEndpointByzantine(ec EndpointCriteria, seq uint64) (bool, error) {
+	bf.Logger.Infof("Checking for BFT Behavior of %s", ec.Endpoint)
+	group := groupCollection{elems: bf.Config.Endpoints}
+	endpointsToProbe, err := group.neighbours(ec, bf.MaxByzantineNodes)
+	if err != nil {
+		bf.Logger.Errorf("groupCollection not configured properly: %v", err)
+		return false, err
+	}
+
+	// Create an attestation source
+	config := bf.Config
+	config.Endpoints = endpointsToProbe
+	attestation_puller := bf.AttestationSourceFactory(config)
+	defer attestation_puller.Close()
+	attestations := attestation_puller.PullAttestations(seq)
+	return bf.ConfirmByzantineBehavior(attestations), nil
+}
+
+func (bf *BlockFetcher) isBlockWitheld(seq uint64) (bool, error) {
+	// this should keep the current endpoint as reference and try to detect if block witholding
+	// takes place by probing (MaxByzantineNodes + 1) endpoints
+	bf.Logger.Info("Checking for Block withholding")
+	malicious, err := bf.isEndpointByzantine(bf.currentEndpoint, seq)
+	if err != nil {
+		return false, err
+	}
+	if malicious {
+		bf.suspects.insert(bf.currentEndpoint.Endpoint)
+		return true, nil
+	}
+	return false, nil
+}
+
+func (bf *BlockFetcher) isEndpointSuspected() bool {
+	return bf.suspects.has(bf.currentEndpoint.Endpoint)
+}
+
+//PullBlock pulls blocks from orderers in a BFT way.
+func (bf *BlockFetcher) PullBlock(seq uint64) *common.Block {
+	retriesLeft := bf.MaxPullBlockRetries
+
+	if bf.suspects.max == 0 {
+		bf.suspects = suspectSet{max: uint(bf.MaxByzantineNodes)}
+	}
+
+	blkSourceOp := CURRENTSOURCE
+	bf.currentSeq = seq
+
+	for {
+
+		if retriesLeft == 0 && bf.MaxPullBlockRetries > 0 {
+			bf.Logger.Errorf("Failed pulling block [%d]: retry count exhausted(%d)", seq, bf.MaxPullBlockRetries)
+			return nil
+		}
+
+		blkSource := bf.getBlockSource(blkSourceOp)
+
+		if blkSource == nil {
+			bf.Logger.Errorf("Pull Block sources exhausted!!!")
+		}
+
+		start_time := bf.TimeNow()
+		block := blkSource.PullBlock(seq)
+
+		if block != nil {
+			return block
+		}
+
+		// endpoint might be unavailable so shuffle
+		if start_time.Add(time.Duration((int64(bf.Config.FetchTimeout) * bf.ShuffleTimeoutThrehold / 100))).After(time.Now()) {
```

Comment:
I don't understand, why would we care how long it took to pull the block? we said we would shuffle on a periodical manner that is independent of how fast or slow blocks are received.

The only thing that affects when the blocks are received are whether they are *not* received in time (after FetchTimeout) and then we check for withholding.

Reply:
@shivdeep-singh-ibm I don't think you addressed this comment

Reply:
This is added to account for cases when node is not available, i.e if we get a nil block before timeout, due to some reason